### PR TITLE
Fix #1636 - Support and optimize constant shifts

### LINKED ISSUE #1636: [FEATURE] shl, bit_shl, shr, bit_shr constant support
I need vshl_n_u64 and friends. I will just write them down by hand for now but ideally we should have them exposed in a normal way.

### PR DESCRIPTION
**Pull request by jfalcou**

### CODE REVIEW

File: include/eve/arch/cpu/wide.hpp [hunk 1]
```diff
@@ -735,13 +735,32 @@ namespace eve
       return detail::self_shl(w, s);
     }
 
+    //! @brief Perform the compound left-shift on all the eve::wide lanes with a constant and assign
+    //! the result to current one.
+    template<typename S, S V>
+    friend EVE_FORCEINLINE auto& operator<<=(wide& w, std::integral_constant<S,V> const& s) noexcept
```
Comment (DenisYaroshevskiy) [inline include/eve/arch/cpu/wide.hpp:741]: Why not eve::index, why constant

File: include/eve/detail/function/simd/arm/neon/bit_compounds.hpp [hunk 2]
```diff
@@ -75,6 +101,33 @@ namespace eve::detail
...
+    else  if constexpr( c == category::int32x4  ) v = vshrq_n_s32(v, S);
+    else  if constexpr( c == category::int16x8  ) v = vshrq_n_s16(v, S);
+    else  if constexpr( c == category::int8x16  ) v = vshrq_n_s8 (v, S);
+    else  if constexpr( c == category::uint64x2 ) v = vshrq_n_u64(v, S);
+    else  if constexpr( c == category::uint32x4 ) v = vshrq_n_u32(v, S);
+    else  if constexpr( c == category::uint16x8 ) v = vshrq_n_u16(v, S);
+    else  if constexpr( c == category::uint8x16 ) v = vshrq_n_u8 (v, S);
+
+    return v;
+  }
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/function/simd/arm/neon/bit_compounds.hpp:129]: sve also has these:
svasrd

**Review by DenisYaroshevskiy — COMMENTED** [review]
Can you also remove all "shift by const" https://github.com/jfalcou/eve/blob/0267b6c12e3c72c05016a95d2066dfc6bf297357/include/eve/detail/shuffle_v2/simd/common/shuffle_l2.hpp#L34

There is another implementation in neon
https://github.com/jfalcou/eve/blob/0267b6c12e3c72c05016a95d2066dfc6bf297357/include/eve/detail/shuffle_v2/simd/arm/neon/shuffle_l2.hpp#L15-L23

File: include/eve/arch/cpu/wide.hpp (hunk 1 above)
Comment (jfalcou) [reply include/eve/arch/cpu/wide.hpp:741]: it's more flexible, index is inheriting from integral constant.
Do we want to keep ppl in eve::index ?

File: include/eve/detail/function/simd/arm/neon/bit_compounds.hpp (hunk 2 above)
Comment (jfalcou) [reply include/eve/detail/function/simd/arm/neon/bit_compounds.hpp:129]: Good catch

**Comment by jfalcou** [comment]
Will remove the old code yup

File: include/eve/arch/cpu/wide.hpp (hunk 1 above)
Comment (DenisYaroshevskiy) [reply include/eve/arch/cpu/wide.hpp:741]: Yeah, index is the correct type here

**Review by DenisYaroshevskiy — APPROVED** [review]

**Comment by DenisYaroshevskiy** [comment]
> Will remove the old code yup

shame you forgot but w/e

**Comment by jfalcou** [comment]
> > Will remove the old code yup
> 
> shame you forgot but w/e

Gonna be done ;)

image

**Pull request merged**